### PR TITLE
Add pull request reviews, pull request review comments and pull request review comments (yes, that's right)

### PR DESCRIPTION
**Pull request by BenJam**
As above. WIP

### CODE REVIEW

**Comment by BenJam** [comment]
uses https://github.com/octobox/octobox/pull/1141 in fixtures to test (so check out that thread for expected behaviour). Does not support commit-level comments.

**Comment by BenJam** [comment]
hmmmm, this is currently storing a lot of 'no description given' comments where there are body-less reviews/comments

**Comment by BenJam** [comment]
test currently seems to be skipping `update_comments` in `self.sync`. Any pointers appreciated if anyone gets a look see.

**Comment by BenJam** [comment]
back at this: looks like this isn't calling download comments because it's not getting a gh client.

**Comment by andrew** [comment]
Comment syncing is still optional and enabled with `INCLUDE_COMMENTS=1` ENV var, so it'll need to be turned on in the tests that should sync comments, at least `test/models/subject_test.rb` and `test/controllers/hooks_controller_test.rb`, with the following code:

```ruby
setup do
  stub_include_comments
end
```

File: test/models/subject_test.rb [hunk 1]
```diff
@@ -228,6 +228,26 @@ class SubjectTest < ActiveSupport::TestCase
...
+    reviews = { status: 200, body: file_fixture('subject_58_reviews.json'), headers: { 'Content-Type' => 'application/json' } }
+    comments_for_review = { status: 200, body: file_fixture('subject_58_comments_for_review.json'), headers: { 'Content-Type' => 'application/json' } }
+    review_comments = { status: 200, body: file_fixture('subject_58_review_comments.json'), headers: { 'Content-Type' => 'application/json' } }
+
+    stub_request(:get, remote_subject["url"]+'/reviews').and_return(reviews)
+    stub_request(:get, remote_subject["url"]+'/comments').and_return(review_comments)
+    stub_request(:get, remote_review["pull_request_url"]+'/reviews/'+remote_review["id"].to_s).and_return(comments_for_review)
+
+    Octobox.expects(:include_comments?).returns(true)
+    Subject.any_instance.expects(:update_comments)
```
Comment (andrew) [inline test/models/subject_test.rb]: This line is actually stubbing out the `update_comments` call, so it doesn't actually run, only asserting that it get's called.

Recommend you remove those two lines and use `stub_include_comments` instead.

**Comment by BenJam** [comment]
ta!

File: app/models/subject.rb [hunk 2]
```diff
@@ -78,7 +78,7 @@ def self.sync(remote_subject)
     return unless subject.persisted?
 
     subject.update_labels(remote_subject['labels']) if remote_subject['labels'].present?
-    subject.update_comments if Octobox.include_comments? && subject.has_comments?
+    subject.update_comments if Octobox.include_comments? && (subject.has_comments? | subject.pull_request?)
```
Comment (andrew) [inline app/models/subject.rb]:
```suggestion
    subject.update_comments if Octobox.include_comments? && (subject.has_comments? || subject.pull_request?)
```

File: app/controllers/hooks_controller.rb [hunk 3]
```diff
@@ -8,7 +8,7 @@ def create
     when 'issues', 'issue_comment'
       SyncSubjectWorker.perform_async_if_configured(payload['issue'])
     when 'pull_request', 'pull_request_review', 'pull_request_review_comment'
-      SyncSubjectWorker.perform_async_if_configured(payload['pull_request'])
```
Comment (andrew) [inline app/controllers/hooks_controller.rb]: I think it'd be good tosplit `'pull_request'` out from the review and review_comment section, so it still calls `SyncSubjectWorker` here

Comment (BenJam) [reply app/controllers/hooks_controller.rb]: will do now

Comment (BenJam) [reply app/controllers/hooks_controller.rb]: I've pushed this but: it'll still sync the comments for the PR as a whole. I could split it out to check for only reviews but I'm not sure there's that much additional work to call the comments on the PR too... Up to you if you want it further split. I'm easy either way.

**Comment by chrisarcand** [comment]
Whoa neat. @BenJam @andrew I don't _think_ anything here conflicts with it, but just as an FYI to keep in mind with this review request area, here's the branch I'm currently testing for #836, if we want to ensure this all lines up nicely : https://github.com/octobox/octobox/compare/master...chrisarcand:add-team-vs-individual-review-requests

**Comment by andrew** [comment]
@chrisarcand that's looking great, this PR only pulls in the review related comments, doesn't touch any requests. That branch is looking great, can't wait for the PR!

**Comment by andrew** [comment]
@BenJam do the review comments visually look any different to regular comments?

**Comment by BenJam** [comment]
it's all one comment stream atm.

**Comment by andrew** [comment]
@BenJam perhaps we can add store the `state` field if it's present (which it only will be for reviews) and then add an icon to the header of the comment.

File: app/models/subject.rb [hunk 4]
```diff
@@ -203,8 +204,10 @@ def download_reviews
     return unless github_client && pull_request?
     reviews = github_client.get(url + '/reviews', since: comments.order('created_at ASC').last.try(:created_at))
     reviews.each do |review|
-        reviews.concat download_comments_for_review(review) if review[:body].empty?
+      if review[:state] == "COMMENTED"
+        reviews.concat download_comments_for_review(review)
```
Comment (andrew) [inline app/models/subject.rb]: I believe that users can leave comments when the review is in any state, not just commented

Comment (BenJam) [reply app/models/subject.rb]: Yeah that’s a mental typo.

**Comment by BenJam** [comment]
@chrisarcand I can't see anything that's going to screw up your PR significantly by merging this. I'd quite like to get it out and move everyone onto comment-threads asap.

**Comment by chrisarcand** [comment]
For sure, go for it, looks great!

**Comment by BenJam** [comment]
Cool, just going to address a little bug in which reviews aren't stored with the correct datetime

**Comment by BenJam** [comment]
Such meta:

Screenshot 2019-06-18 at 13 51 12

**Comment by bertie** [comment]
Hey!

**Pull request merged**